### PR TITLE
feat(ge): GE buy/sell window contextual trade data (#665)

### PR DESCRIPTION
**Pull request by mablack01**
## Summary

Replaces the stock item description in the GE buy/sell **SETUP** panel and the in-flight offer **DETAILS** panel with contextual FlipSmart trade data. The buy panel shows Daily Volume, Buy Limit, and Wiki insta-buy price. The sell panel shows Breakeven, Tax Applied, and Your Profit — all three recalculate live as the player adjusts price or quantity. Profit is color-coded green/red/white. The same data surfaces on the DETAILS slot-status panel after an offer is placed, updated per-frame via \`BeforeRender\` with a \`setText\` short-circuit when the text is already current.

**Backend dependency:** Requires \`GET /items/{itemId}/daily-volume\` (Flip-Smart/flip-smart#672) to be merged and deployed. The plugin compiles and runs without it — the buy line will show \`Daily Volume: N/A\` until the endpoint is live. As of this PR, the endpoint is live in production.

### New Files

- **\`GeOfferDescriptionFormatter.java\`** — pure-function formatters for buy/sell descriptions, breakeven, tax, profit, and color tagging. Zero RuneLite dependencies; fully unit-testable. Houses the GE tax constants (2%, 5M cap, 50gp exempt threshold).
- **\`GeOfferDescriptionService.java\`** — wires the formatter to RuneLite. Handles SETUP panels via \`ScriptCallbackEvent\` (\`geBuyExamineText\`/\`geSellExamineText\`) and DETAILS panels via \`BeforeRender\`. Subscribes to \`MenuOptionClicked\` to record which slot the user actually clicked, avoiding the off-by-one caused by relying solely on the \`GE_SELECTEDSLOT\` varbit.
- **\`BuyPriceLookup.java\`** — shared utility resolving the player's recorded average buy price from active flips. Used by both the sell-description formatter and \`GrandExchangeSlotOverlay\`.
- **\`GeOfferDescriptionFormatterTest.java\`** — 23 unit tests covering all formatting rules and edge cases (tax cap, exempt threshold, color tagging, sign prefix, k/M shorthand, breakeven minimality, null buy-price fallbacks).

### Modified Files

- **\`FlipSmartPlugin.java\`** — injects \`GeOfferDescriptionService\`; routes \`ScriptCallbackEvent\` and \`MenuOptionClicked\` to it.
- **\`FlipSmartApiClient.java\`** — adds \`getDailyVolumeAsync(itemId)\` calling \`GET /items/{itemId}/daily-volume\` with a 5-minute \`ConcurrentHashMap\` cache and a non-blocking \`peekCachedDailyVolume\` for the script-callback hot path.
- **\`GrandExchangeSlotOverlay.java\`** — delegates \`getBuyPriceForItem\` to \`BuyPriceLookup\`. No behavioral change to the existing slot-hover P&L tooltip.

## Key Implementation Notes

**Slot identification.** The \`GE_SELECTEDSLOT\` varbit tracks the hovered tile, not the panel that opened. Relying on it caused data from slot N to render in slot N+1's panel. Fixed by subscribing to \`MenuOptionClicked\` and recording the slot the user explicitly clicked; the varbit is used only as a cold-start fallback.

**Text alignment.** The stock GE description widget is CENTER-aligned for a single-line string. Forced LEFT alignment so our 3-line replacement renders flush left. The \`DETAILS_DESC\` widget also sits 10px further left than \`SETUP_DESC\` in Jagex's hand-tuned layout, which exposed a parent-background strip; the fix shifts \`DETAILS_DESC\` left 10px and widens it to compensate.

**Color choice.** Loss-red is \`#ff4040\` rather than pure \`#ff0000\`. The GE description panel has a stone/beige background and pure red reads poorly at widget font sizes.

### Automated Tests

- 23 new unit tests in \`GeOfferDescriptionFormatterTest\` — all pass.
- 24 pre-existing tests (PlayerSession, MotdService, ExamplePlugin) — all pass.
- Total: 47 tests, 0 failures.
- Build: \`./gradlew compileJava test\` — BUILD SUCCESSFUL.

### Manual Test Plan

**Buy SETUP panel**
- [x] Open a buy offer for any item. Confirm three lines render: \`Daily Volume: N,NNN\`, \`Buy limit: N / 4h\`, \`Wiki insta-buy: Ngp\`. Text is left-aligned, not centered.
- [x] Open a buy offer while the backend endpoint is unreachable. Confirm \`Daily Volume: N/A\` — plugin does not crash.

**Sell SETUP panel**
- [x] Open a sell offer for an item with active-flip buy history in FlipSmart. Confirm: \`Breakeven: X gp\`, \`Tax applied: X gp\`, \`Your profit: +X gp\` with correct values.
- [x] Adjust the listed sell price using +/- buttons and via chatbox price entry. Confirm \`Tax applied\` and \`Your profit\` recalculate live on each change, and that the profit line color flips green/red/white at the breakeven crossover.
- [x] Open a sell offer for an item with no buy history. Confirm \`Breakeven: ?\` and \`Your profit: ?\` (white text).

**In-flight DETAILS panel**
- [x] Place a buy offer. Open its slot status panel. Confirm the same three buy lines appear in the description area, left-aligned.
- [x] Place a sell offer. Open its slot status panel. Confirm the same three sell lines appear with correct values for the filled price.

**Slot off-by-one regression**
- [x] Open the GE with multiple active offers. Click slot 1, then slot 2, then slot 3 in sequence. Confirm the description data on each panel matches that slot's item — not the previous slot's item.

**Cleanup**
- [x] Close the GE while an offer is open. Reopen to a different item. Confirm no residual FlipSmart text.
- [x] Disable the FlipSmart plugin. Reopen a GE offer. Confirm vanilla examine text returns.
- [x] Confirm the GE history screen, collection box, and search results are unaffected.

## Out of Scope

- GE history screen, collection box, GE search results — untouched by design.
- Buy-window daily volume does not update dynamically (fetched once on window open, cached 5 min).
- Overlay approaches — not used (no \`Overlay\` or \`drawAfterInterface\` in any new file).

## Deployment Notes

- Backend endpoint \`GET /items/{itemId}/daily-volume\` is already live in production (Flip-Smart/flip-smart#672).
- No new plugin config options or settings UI changes.
- No plugin-hub metadata changes.

## Checklist

- [x] All 9 ACs addressed; AC7/AC8/AC9 satisfied by construction
- [x] 47 tests pass (\`./gradlew compileJava test\`)
- [x] No paint overlay used — no \`Overlay\` in any new file
- [x] \`Thread.currentThread().interrupt()\` NOT called on executor-managed threads
- [x] Thread safety: daily-volume cache uses \`ConcurrentHashMap\`; \`peekCachedDailyVolume\` is non-blocking
- [x] No \`.claude/\`, \`CLAUDE.md\`, or Claude config added to this public repo

## Related Issues

Closes Flip-Smart/flip-smart#665

Backend PR (already merged): Flip-Smart/flip-smart#672

---

### Codacy Quality Gate — fixed in this PR
- \`f191534\` Lizard_ccn-medium / Lizard_nloc-medium / PMD_NPathComplexity GeOfferDescriptionService.java:172 — extracted \`resolveOfferContext\` + \`resolveSlotContext\` from \`onBeforeRender\`
- \`f191534\` Lizard_ccn-medium GeOfferDescriptionService.java:111 — extracted \`isDetailsContainerVisible\` + \`writeToObjectStack\` from \`handleExamine\`
- \`67c7721\` Lizard_ccn-medium GeOfferDescriptionService.java:238 — extracted \`resolveActiveSlot\` from \`resolveSlotContext\` to move slot-fallback branching into a dedicated helper

### Codacy AI Reviewer — deferred (in-thread rationale)
- GeOfferDescriptionService.java:288 — redundant in-flight fetch: TTL bounds fanout to at most one per cache miss per item; deduplication tracking is out of scope for this PR
- GeOfferDescriptionFormatter.java:183 — long overflow in \`calculateBreakevenPrice\`: \`recordedBuyPrice\` is bounded by the OSRS GE per-item cap well below \`Integer.MAX_VALUE\`; risk is negligible in practice

### Codacy AI Reviewer — false positives (in-thread rationale)
- GeOfferDescriptionFormatter.java:24 — GE_TAX_RATE 0.02 is correct (OSRS raised the tax to 2% in November 2024)
- FlipSmartApiClient.java:2521 — \`peekCachedDailyVolume\` intentionally returns \`null\` to signal "not cached/expired"; the caller guards against null before chaining

### CODE REVIEW

File: src/main/java/com/flipsmart/GeOfferDescriptionFormatter.java [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+		{
+			return recordedBuyPrice;
+		}
+		// Smallest sell price S such that S - floor(S * 0.02) >= recordedBuyPrice.
+		// 1) Start from the closed-form estimate (overshoots by 1 due to ceiling).
+		// 2) Walk up if the cap region or floor-truncation leaves us short.
+		// 3) Walk down to guarantee minimality (the closed form typically yields
+		//    S=103 for buy=100 when S=102 already satisfies — without this step
+		//    we'd over-quote the breakeven by one gp).
+		int candidate = (int) Math.ceil(recordedBuyPrice / (1.0 - GE_TAX_RATE));
```
[the remark this answers is not in the document]
Comment (mablack01) [reply src/main/java/com/flipsmart/GeOfferDescriptionFormatter.java:183]: Deferred. `recordedBuyPrice` is a player's recorded average buy price in gp; the OSRS GE has a hard per-item cap below 2.147B (`Integer.MAX_VALUE`), and the closed-form `recordedBuyPrice / 0.98` only overflows when the input is already within ~2% of `Integer.MAX_VALUE` — values the live game cannot produce. The walk-up/walk-down loop operates on `int candidate` and would also need lifting, expanding the surface. Leaving as int to keep the algorithm tight; will revisit if OSRS ever raises the per-item cap.

File: src/main/java/com/flipsmart/GeOfferDescriptionFormatter.java [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+{
+	// Color codes used inline in the returned RuneScript strings.
+	// Format is the standard widget hex tag, e.g. <col=00ff00>green</col>.
+	static final String COLOR_GREEN = "00ff00";
+	static final String COLOR_RED = "ff4040";
+	static final String COLOR_WHITE = "ffffff";
+	static final String COLOR_LABEL = "ffb83f";  // GE description "highlight" amber
+
+	private static final int GE_TAX_EXEMPT_THRESHOLD = 50;
+	private static final double GE_TAX_RATE = 0.02;
```
[the remark this answers is not in the document]
Comment (mablack01) [reply src/main/java/com/flipsmart/GeOfferDescriptionFormatter.java:24]: False positive. The OSRS Grand Exchange tax has been 2% since the November 2024 rate increase (raised from 1%). 0.02 is correct; the same constant is enforced backend-side.

File: src/main/java/com/flipsmart/GeOfferDescriptionService.java [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+	}
+
+	private String buildBuyDescription(int itemId)
+	{
+		Integer dailyVolume = getCachedDailyVolume(itemId);
+		// Kick off async fetch — once the cache is warm, subsequent rebuilds of
+		// the description (same item or other items) will pick up the value.
+		if (dailyVolume == null)
+		{
+			apiClient.getDailyVolumeAsync(itemId).whenComplete((v, ex) ->
```
[the remark this answers is not in the document]
Comment (mablack01) [reply src/main/java/com/flipsmart/GeOfferDescriptionService.java:328]: Deferred. The 5-minute cache TTL (`DAILY_VOLUME_CACHE_TTL_MS`) bounds the redundant-request window to at most one fanout per item per cache miss — the GE buy window only fires `geBuyExamineText` a handful of times per item selection, and once any in-flight request completes the cache absorbs subsequent rebuilds. Adding an in-flight map would harden it but isn't worth the extra state for the current event cadence. Tracked separately if we ever see this in metrics.

File: src/main/java/com/flipsmart/FlipSmartApiClient.java [hunk 4]
```diff
@@ -2470,4 +2470,71 @@ public CompletableFuture<MotdResponse> getMotdAsync()
...
+	 * Non-blocking peek into the daily-volume cache. Returns a completed future
+	 * with the cached value when fresh, or {@code null} when not cached / expired.
+	 * Safe to call from the RuneLite client thread (no network I/O, no locks).
+	 */
+	public CompletableFuture<Integer> peekCachedDailyVolume(int itemId)
+	{
+		CachedDailyVolume cached = dailyVolumeCache.get(itemId);
+		if (cached == null || cached.isExpired())
+		{
+			return null;
```
[the remark this answers is not in the document]
Comment (mablack01) [reply src/main/java/com/flipsmart/FlipSmartApiClient.java:2521]: False positive in context. `peekCachedDailyVolume` is the *non-blocking peek* variant — its contract is specifically to return `null` to signal "not cached / expired" so the caller can decide whether to fall back to `getDailyVolumeAsync`. Returning `CompletableFuture.completedFuture(null)` would collapse the "not cached" and "cached as null" cases (the latter is a valid state — the API returned no volume data for the item). The only call site in `GeOfferDescriptionService.getCachedDailyVolume` null-checks the future before calling `.get()`.

File: src/main/java/com/flipsmart/GeOfferDescriptionFormatter.java [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+	{
+		long abs = Math.abs(value);
+		String sign = value < 0 ? "-" : "";
+		if (abs >= 1_000_000)
+		{
+			return sign + stripTrailingZero(String.format("%.1f", abs / 1_000_000.0)) + "M";
+		}
+		if (abs >= 1_000)
+		{
+			return sign + stripTrailingZero(String.format("%.1f", abs / 1_000.0)) + "k";
```
[the remark this answers is not in the document]
Comment (mablack01) [reply src/main/java/com/flipsmart/GeOfferDescriptionFormatter.java]: Fixed in d6a42e4 — `Locale.ROOT` added to both `String.format("%.1f", ...)` calls in `formatShortLower` (lines 232 and 236 in the current file).

File: src/main/java/com/flipsmart/GeOfferDescriptionFormatter.java [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+	 * Lowercase k/M shorthand matching the AC examples ({@code 100k}, {@code 1.5M}).
+	 * Distinct from {@link GpUtils#formatGP} which uses uppercase K.
+	 */
+	static String formatShortLower(long value)
+	{
+		long abs = Math.abs(value);
+		String sign = value < 0 ? "-" : "";
+		if (abs >= 1_000_000)
+		{
+			return sign + stripTrailingZero(String.format("%.1f", abs / 1_000_000.0)) + "M";
```
[the remark this answers is not in the document]
Comment (mablack01) [reply src/main/java/com/flipsmart/GeOfferDescriptionFormatter.java]: Fixed in d6a42e4 — same change as the sibling thread; `Locale.ROOT` applied to both `%.1f` format calls in `formatShortLower`.

File: src/main/java/com/flipsmart/GeOfferDescriptionFormatter.java [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+
+	private static String colorTag(String hex)
+	{
+		return "<col=" + hex + ">";
+	}
+
+	/** Comma-separated exact gp count: {@code 1234567} -> {@code "1,234,567"}. */
+	static String formatExact(long value)
+	{
+		return String.format("%,d", value);
```
[the remark this answers is not in the document]
Comment (mablack01) [reply src/main/java/com/flipsmart/GeOfferDescriptionFormatter.java]: Fixed in 7c68cda — `Locale.ROOT` added to `String.format("%,d", value)` in `formatExact`. The `%,d` grouping separator is locale-sensitive (e.g. `.` in de_DE, space in fr_FR); pinning to `Locale.ROOT` keeps the comma separator that the AC examples assume.

File: src/main/java/com/flipsmart/GeOfferDescriptionFormatter.java [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+		long totalProfit = (long) profitPerItem * (long) Math.max(quantity, 1);
+		String color = colorForProfit(totalProfit);
+		String sign = totalProfit > 0 ? "+" : "";
+
+		StringBuilder sb = new StringBuilder(label);
+		sb.append(colorTag(color));
+		sb.append(sign).append(formatExact(totalProfit)).append("gp");
+		if (quantity > 1)
+		{
+			String perItemSign = profitPerItem > 0 ? "+" : "";
```
[the remark this answers is not in the document]
Comment (mablack01) [reply src/main/java/com/flipsmart/GeOfferDescriptionFormatter.java:164]: Intentional. The total profit already renders with a leading `+` for positive values; the per-item summary inside the parenthetical mirrors that convention for visual consistency at a glance. AC5 didn't forbid it — leaving as-is.
